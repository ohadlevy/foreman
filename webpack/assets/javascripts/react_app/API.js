import $ from 'jquery';

export default {
  get(url) {
    $.ajaxPrefilter((options, originalOptions, jqXHR) => {
      jqXHR.originalRequestOptions = originalOptions;
    });
    return $.getJSON(url);
  },
  post(url, data) {
    return $.ajax({
      url,
      contentType: 'application/json',
      type: 'post',
      dataType: 'json',
      data
    });
  },
  markNotificationAsRead(id) {
    const data = JSON.stringify({ seen: true });

    $.ajax({
      url: `/notification_recipients/${id}`,
      contentType: 'application/json',
      type: 'put',
      dataType: 'json',
      data,
      error(jqXHR, textStatus, errorThrown) {
        /* eslint-disable no-console */
        console.log(jqXHR);
      }
    });
  },
  markGroupNotificationAsRead(group) {
    $.ajax({
      url: `/notification_recipients/group/${group}`,
      contentType: 'application/json',
      type: 'PUT',
      error(jqXHR, textStatus, errorThrown) {
        /* eslint-disable no-console */
        console.log(jqXHR);
      }
    });
  }
};
