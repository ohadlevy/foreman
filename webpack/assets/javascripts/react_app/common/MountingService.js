import React from 'react';
import StatisticsChartsList from '../components/charts/StatisticsChartsList';
import FlashNotifications from '../components/notifications/FlashNotifications';

import ReactDOM from 'react-dom';

export function mount(component, selector, data) {

  const components = {
    StatisticsChartsList: {
      type: StatisticsChartsList,
      markup: <StatisticsChartsList data={data}/>
    },
    FlashNotifications: {
     type: FlashNotifications,
     markup: <FlashNotifications flash={data}/>
   }
  };

  const reactNode = document.querySelector(selector);

  if (reactNode) {
    ReactDOM.render(components[component].markup, reactNode);
  } else {
    const componentName = components[component].type.name;

    // eslint-disable-next-line no-console
    console.log(`Cannot find \'${selector}\' element for mounting the \'${componentName}\'`);
  }
}
