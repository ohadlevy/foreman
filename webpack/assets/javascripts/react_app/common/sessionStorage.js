import Immutable from 'seamless-immutable';

export const loadState = () => {
  try {
    const seralizedState = sessionStorage.getItem('state');

    if (seralizedState === null) {
      return undefined;
    }
    return Immutable(JSON.parse(seralizedState));
  } catch (err) {
    return undefined;
  }
};

export const saveState = state => {
  try {
    const seralizedState = JSON.stringify(state);

    sessionStorage.setItem('state', seralizedState);
  } catch (err) {
    // ignore errors, todo log somewhere
  }
};
