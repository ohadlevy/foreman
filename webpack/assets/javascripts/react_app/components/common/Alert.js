import React from 'react';
import Icon from './Icon';
import { ALERT_CSS } from '../../constants';

/* eslint-disable max-params */
const Alert = ({children, type = 'info', dismissable = false, css = null}) => {
  let classNames = css ? ALERT_CSS[type] + ' ' + css : ALERT_CSS[type];

  if (dismissable) {
    classNames += ' ' + ALERT_CSS.dismissable;
  }

  const closeBtn = (
    <button className="close" data-dismiss="alert" aria-hidden="true">
      <Icon type="close" />
    </button>
  );

  return (
    <div className={classNames}>
      {dismissable && closeBtn}
      {children}
    </div>
  );
};

export default Alert;
