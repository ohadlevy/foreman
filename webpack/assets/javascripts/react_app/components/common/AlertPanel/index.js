import React from 'react';
import getAlertClass from './Alert.consts';
import Icon from '../Icon/';
import Button from '../forms/Button';

export default ({
  className = '',
  type,
  children,
  title,
  onClose
}) => {
  const CloseButton = ({onClick}) => (
    <Button className="close" aria-hidden="true" onClick={onClick}>
      <Icon type="close" />
    </Button>
  );

  return (
    <div
      className={`${getAlertClass(type, onClose)}${className ? ' ' + className : ''}`}
    >
      {onClose && <CloseButton onClick={onClose} />}
      <Icon type={type} />
      {title && <strong>{title}</strong>}
      {title && <br />}
      {children}
    </div>
  );
};
