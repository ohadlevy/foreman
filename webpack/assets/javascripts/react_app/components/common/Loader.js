import React from 'react';
import styles from './LoaderStyles';
import {STATUS} from '../../constants';
import MessageBox from './MessageBox';

const Loader = ({ status, children }) => {
let content;

  switch (status) {
    case STATUS.PENDING: {
      content = <div className="spinner spinner-lg"></div>;
      break;
    }
    case STATUS.RESOLVED: {
      content = children[0];
      break;
    }
    case STATUS.ERROR: {
      content = children[1];
      break;
    }
    default:
      content = (<MessageBox icontype="error-circle-o" />);
      break;
  }

  return (
    <div style={styles.root}>
      {content}
    </div>
  );
};

export default Loader;
