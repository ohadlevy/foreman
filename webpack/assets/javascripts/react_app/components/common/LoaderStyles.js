export default {
  root: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    flexGrow: 1
  }
};
