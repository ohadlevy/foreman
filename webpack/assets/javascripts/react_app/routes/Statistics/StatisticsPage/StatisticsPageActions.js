import API from '../../../API';

import {
  STATISTICS_PAGE_META_RESOLVED,
  STATISTICS_PAGE_META_FAILED,
  STATISTICS_PAGE_HIDE_LOADING,
  STATISTICS_PAGE_URL,
  STATISTICS_DATA_REQUEST,
  STATISTICS_DATA_SUCCESS,
  STATISTICS_DATA_FAILURE,
} from '../constants';
import { ajaxRequestAction } from '../../../redux/actions/common';

export const getStatisticsMeta = (
  url = STATISTICS_PAGE_URL
) => async dispatch => {
  const onFetchSuccess = ({ data }) => {
    dispatch(hideLoading());
    dispatch({
      type: STATISTICS_PAGE_META_RESOLVED,
      payload: data,
    });
  };

  const onFetchError = ({ message }) => {
    dispatch(hideLoading());
    dispatch({
      type: STATISTICS_PAGE_META_FAILED,
      payload: {
        hasError: true,
        message: {
          type: 'error',
          text: message,
        },
      },
    });
  };
  try {
    const response = await API.get(url);
    return onFetchSuccess(response);
  } catch (error) {
    return onFetchError(error);
  }
};

export const getStatisticsData = charts => dispatch =>
  Promise.all(
    charts.map(chart =>
      ajaxRequestAction({
        dispatch,
        requestAction: STATISTICS_DATA_REQUEST,
        successAction: STATISTICS_DATA_SUCCESS,
        failedAction: STATISTICS_DATA_FAILURE,
        url: chart.url,
        item: chart,
      })
    )
  );

const hideLoading = () => ({
  type: STATISTICS_PAGE_HIDE_LOADING,
});
