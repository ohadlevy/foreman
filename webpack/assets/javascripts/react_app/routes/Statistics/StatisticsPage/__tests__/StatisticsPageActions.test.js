import API from '../../../../API';

import { testActionSnapshotWithFixtures } from '../../../../common/testHelpers';
import { getStatisticsMeta, getStatisticsData } from '../StatisticsPageActions';
import {
  statisticsProps,
  successRequestData,
} from '../StatisticsPage.fixtures';

jest.mock('../../../../API');

const runStatisticsAction = (callback, props, serverMock) => {
  API.get.mockImplementation(serverMock);

  return callback(props);
};

const fixtures = {
  'should fetch statisticsMeta': () =>
    runStatisticsAction(getStatisticsMeta, {}, async () => ({
      data: statisticsProps.statistics,
    })),
  'should fetch statisticsMeta and fail': () =>
    runStatisticsAction(getStatisticsMeta, {}, async () => {
      throw new Error('some-error');
    }),
  'should fetch statisticsData': () =>
    runStatisticsAction(getStatisticsData, successRequestData, async () => ({
      data: statisticsProps.charts,
    })),
  'should fetch statisticsData and fail': () =>
    runStatisticsAction(getStatisticsData, successRequestData, async () => {
      throw new Error('some-error');
    }),
};

describe('StatisticsPage actions', () =>
  testActionSnapshotWithFixtures(fixtures));
