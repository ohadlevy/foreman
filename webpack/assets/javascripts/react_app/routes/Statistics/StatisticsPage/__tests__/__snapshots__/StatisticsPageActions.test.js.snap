// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticsPage actions should fetch statisticsData 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "id": "operatingsystem",
        "search": "/hosts?search=os_title=~VAL~",
        "title": "OS Distribution",
        "url": "statistics/operatingsystem",
      },
      "type": "STATISTICS_DATA_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "id": "architecture",
        "search": "/hosts?search=facts.architecture=~VAL~",
        "title": "Architecture Distribution",
        "url": "statistics/architecture",
      },
      "type": "STATISTICS_DATA_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "0": Object {
          "data": Array [
            Array [
              "centOS 7.1",
              6,
            ],
          ],
          "id": "operatingsystem",
          "search": "/hosts?search=os_title=~VAL~",
          "title": "OS Distribution",
          "url": "statistics/operatingsystem",
        },
        "id": "operatingsystem",
        "search": "/hosts?search=os_title=~VAL~",
        "title": "OS Distribution",
        "url": "statistics/operatingsystem",
      },
      "type": "STATISTICS_DATA_SUCCESS",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "0": Object {
          "data": Array [
            Array [
              "centOS 7.1",
              6,
            ],
          ],
          "id": "operatingsystem",
          "search": "/hosts?search=os_title=~VAL~",
          "title": "OS Distribution",
          "url": "statistics/operatingsystem",
        },
        "id": "architecture",
        "search": "/hosts?search=facts.architecture=~VAL~",
        "title": "Architecture Distribution",
        "url": "statistics/architecture",
      },
      "type": "STATISTICS_DATA_SUCCESS",
    },
  ],
]
`;

exports[`StatisticsPage actions should fetch statisticsData and fail 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "id": "operatingsystem",
        "search": "/hosts?search=os_title=~VAL~",
        "title": "OS Distribution",
        "url": "statistics/operatingsystem",
      },
      "type": "STATISTICS_DATA_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "id": "architecture",
        "search": "/hosts?search=facts.architecture=~VAL~",
        "title": "Architecture Distribution",
        "url": "statistics/architecture",
      },
      "type": "STATISTICS_DATA_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "error": [Error: some-error],
        "item": Object {
          "id": "operatingsystem",
          "search": "/hosts?search=os_title=~VAL~",
          "title": "OS Distribution",
          "url": "statistics/operatingsystem",
        },
      },
      "type": "STATISTICS_DATA_FAILURE",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "error": [Error: some-error],
        "item": Object {
          "id": "architecture",
          "search": "/hosts?search=facts.architecture=~VAL~",
          "title": "Architecture Distribution",
          "url": "statistics/architecture",
        },
      },
      "type": "STATISTICS_DATA_FAILURE",
    },
  ],
]
`;

exports[`StatisticsPage actions should fetch statisticsMeta 1`] = `
Array [
  Array [
    Object {
      "type": "STATISTICS_PAGE_HIDE_LOADING",
    },
  ],
  Array [
    Object {
      "payload": undefined,
      "type": "STATISTICS_PAGE_META_RESOLVED",
    },
  ],
]
`;

exports[`StatisticsPage actions should fetch statisticsMeta and fail 1`] = `
Array [
  Array [
    Object {
      "type": "STATISTICS_PAGE_HIDE_LOADING",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "hasError": true,
        "message": Object {
          "text": "some-error",
          "type": "error",
        },
      },
      "type": "STATISTICS_PAGE_FETCH_FAILED",
    },
  ],
]
`;
